aggregator_core/src/datastore.rs

```diff
@@ -1069,14 +1069,22 @@ impl<C: Clock> Transaction<'_, C> {
                 "SELECT
                     (SELECT COUNT(*) FROM tasks WHERE task_id = $1) AS task_count,
                     (SELECT COUNT(*) FROM client_reports
-                     WHERE task_id = (SELECT id FROM tasks WHERE task_id = $1)) AS report_count,
+                     JOIN tasks ON tasks.id = client_reports.task_id
+                     WHERE tasks.task_id = $1
+                       AND client_reports.client_timestamp >= COALESCE($2::TIMESTAMP - tasks.report_expiry_age * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)) AS report_count,
```

Comment:
> So in english, we select non-expired reports, unless the task does not provide a report expiry age, in which case we select all reports of the task. Correct?

That's correct. The "optional" nature of `report_expiry_age` behavior is recorded in a few places, though in the task definition itself it's only "documented" as far as being an `Option<Duration>` -- I'll augment documentation there in the next PR in the chain.


> I see in the schema we do deliberately leave the possibility of report expiry being null, to allow for disabling the GC. What cases would this be for?

I think Divvi Up the product will always set a non-NULL report expiry age. Folks might deploy this not wanting expiration at all (or wanting to experiment without expiration), if the data is considered very important, etc. The use cases are a little slim, but I guess my strongest argument is that I don't see a reason to _require_ a GC age.


> I would have imagined the report expiry would be consistent with the task expiry.

I think the two parameters would likely differ. `report_expiry_age` is effectively "how long can data persist before it can be GC'ed" (or, to an outside observer, "how long do collections persist before we can no longer collect them"). `task_expiry_age` is more for "how long until we think we'll no longer need this task" -- at which point new uploads fail.


> I'm also tempted to suggest bringing this statement out into a postgres function, but not sure if that's really a win.

I am admittedly less experienced with Postgres' function-definition capabilities, but my understanding is that the function would execute one or more (independent) statements -- i.e. it seems there may be some performance overhead. What do folks think?

Reply:
Looked into a bit more. The way I was thinking it would be written would be along something to the tune of of
```sql
create function is_expired(client_timestamp TIMESTAMP, report_expiry_age BIGINT) 
returns boolean 
language sql
    return client_timestamp >= COALESCE (now()::timestamp - report_expiry_age * '1 second'::INTERVAL, '-infinity'::TIMESTAMP);
```
But this is at best just syntactic sugar, not really worth pursuing IMO.